lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,249 @@
+        ["authorityKeyIdentifier", "keyid:always", false]
+      ]
+      DEFAULT_SIGNING_DIGEST = OpenSSL::Digest::SHA256.new
+
+      SUMMARY = "Generate a root and intermediate signing CA for Puppet Server"
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca generate [--help|--version]
+  puppetserver ca generate [--config PATH]
+
+Description:
+Generate a root and intermediate signing CA for Puppet Server
+and store generated CA keys, certs, and crls on disk.
+
+To determine the target location, the default puppet.conf
+is consulted for custom values. If using a custom puppet.conf
+provide it with the --config flag
+Options:
+BANNER
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        # Validate config_path provided
+        config_path = input['config']
+        errors = Puppetserver::Utils::FileUtilities.validate_file_paths([config_path])
+        log_errors_and_return(errors) if errors.any?
+
+        # Load, resolve, and validate puppet config settings
+        puppet = PuppetConfig.parse(config_path)
+        log_errors_and_return(puppet.errors) if puppet.errors.any?
+
+        # Generate root and intermediate ca and put all the certificates, crls,
+        # and keys where they should go.
+        generate_root_and_intermediate_ca(puppet)
+
+        # Puppet's internal CA expects these file to exist.
+        Puppetserver::Utils::FileUtilities.ensure_file(puppet.settings[:serial], "001", 0640)
+        Puppetserver::Utils::FileUtilities.ensure_file(puppet.settings[:cert_inventory], "", 0640)
+
+        @logger.inform "Generation succeeded. Find your files in #{puppet.settings[:cadir]}"
+        return 0
+      end
+
+      def log_errors_and_return(errors)
+        errors = Array(errors).compact
+        @logger.err "Error:"
+        errors.each do |message|
+          @logger.err "    #{message}"
+        end
+        return 1
+      end
+
+      def generate_root_and_intermediate_ca(puppet)
+        private_key_length = puppet.settings[:keylength]
+        root_ca_name = puppet.settings[:root_ca_name]
+        int_ca_name = puppet.settings[:ca_name]
+
+        ttl_setting = Puppetserver::Settings::TTLSetting.new(:ca_ttl, puppet.settings[:ca_ttl])
+        log_errors_and_return(ttl_setting.errors) if ttl_setting.errors.any?
+        ca_ttl = ttl_setting.munged_value
+
+        root_key = create_private_key(private_key_length)
+        root_cert = self_signed_ca(root_key, root_ca_name, ca_ttl)
+        root_crl = create_crl_for(root_cert, root_key, ca_ttl)
+
+        int_key = create_private_key(private_key_length)
+        int_csr = create_csr(int_key, int_ca_name)
+        int_cert = sign(root_key, root_cert, int_csr, ca_ttl)
+        int_crl = create_crl_for(int_cert, int_key, ca_ttl)
+
+        Puppetserver::Utils::FileUtilities.ensure_dir(puppet.settings[:cadir])
+
+        future_files = [ { location: puppet.settings[:cacert], content: [int_cert, root_cert]},
```

Comment:
Since we don't use future_files anywhere else we could iterate over the array of them directly, don't know if that's less readable though?

Reply:
I'll play around with it and see what I can do.